package xbe.messages;

message XbeMessage {
    required Header header = 1;

    optional Error error = 2;

    optional Execute execute = 3;
    optional ExecuteAck execute_ack = 4;
    optional ExecuteNak execute_nak = 5;

    optional StatusReq status_req = 6;
    optional Status status = 7;

    optional Finished finished = 8;
    optional FinishedAck finished_ack = 9;

    optional Failed failed = 10;
    optional FailedAck failed_ack = 11;

    optional Shutdown shutdown = 12;
    optional ShutdownAck shutdown_ack = 13;

    optional Terminate terminate = 14;
    optional TerminateAck terminate_ack = 15;

    optional LifeSign life_sign = 16;
}

enum ErrorCode {
    UNKNOWN_ERROR = 0;
}

enum ExecuteNakReason {
    UNKNOWN_REASON = 0;
    RESOURCE_BUSY = 1;
}

enum StatusCode {
    IDLE = 0;
    RUNNING = 1;
    FINISHED = 2;
    FAILED = 3;
}

enum FailReason {
    UNKNOWN = 0;
}

message Header {
    // each message must contain a UUID indentifying the conversation
    required bytes conversation_id = 1;
}

message Error {
    required ErrorCode code = 1;
    optional string message = 2;
}

message Task {
    required string executable = 1;
    repeated string argument = 2;

    message Env {
        required string key = 1;
        required string val = 2;
    }
    repeated Env env = 3;

    optional string stdin = 6;
    optional string stdout = 7;
    optional string stderr = 8;
    optional string wd = 9 [default = "/"]; // working dir
}

message Execute {
    optional Task main_task = 1;
    optional Task status_task = 2;
}

message ExecuteAck {
    optional int32 task = 1;
}

message ExecuteNak {
    required ExecuteNakReason reason = 1;
    optional string message = 2;
}

message StatusReq {
    optional bool execute_status_task = 1;
}

message Status {
    required StatusCode status = 1;
    optional int32 status_task_exit_code = 2;
    optional bytes stdout = 3;
    optional bytes stderr = 4;    
}

message Finished {
    required int32 exitcode = 1;
    optional int32 task = 2;
}

message FinishedAck {
    optional int32 task = 1;
}

message Failed {
    optional int32 task = 1;
    optional FailReason reason = 2;
}

message FailedAck {
    optional int32 task = 1;
}

message Shutdown {
    optional string reason = 1;
}

message ShutdownAck {
}

message Terminate {
    optional int32 task = 1;
}

message TerminateAck {
    optional int32 task = 1;
}

message LifeSign {
    required uint64 tstamp = 1;
    optional StatusCode status = 2;
}

